### ISSUE TITLE
[misc] Support cursors in LineAppender

### ISSUE THREAD

**Issue by k-ye** [taichi-dev/taichi#1537]
**Concisely describe the proposed feature**

`LineAppender` (LA) is used for src2src codegens like GL and Metal backends: https://github.com/taichi-dev/taichi/blob/341566979eee399711b598cee2a7cb34600a6de2/taichi/util/line_appender.h#L10-L15

Currently, it only acts as a queue and supports appending linearly. This can be too restrictive: sometimes the codegen needs to process the entire AST before it can produce some code at an earlier point, resulting in either unnecessary code being emitted, or workarounds to concatenate the lines of multiple `LA`s.

**Describe the solution you'd like (if any)**

I'd like to add a `Cursor` concept to `LA`. It allows us to preserve a cursor, keep appending to the LA, then rewind to that cursor after processing the AST.

For example, this is the added interface:

```cpp
class LineAppender {
   struct Cursor { ... };
   // Returns a cursor that points at the current tail of the LineAppender.
  Cursor make_cursor();

  // Rewinds the LineAppende to where |c| points to.
  //
  // After calling this, |c| will be invalid. I.e. a cursor cannot be fed back
  // more than once.
  void rewind_to_cursor(const Cursor &rc);
};
```

Example usage:

```cpp
LineAppender l;
l.append("a");  // "a"
l.append("b");  // "ab"
auto c = l.make_cursor();
l.append("e");  // "abe"
l.rewind_to_cursor(c);  // |c| is invalid after this call
l.append("c");  // "abce"
l.append("d");  // "abcde"
l.lines();      // "abcde"
```

@archibate thoughts?

**Comment by archibate** [comment]
FYI I'm using two LineAppender in the C backend: `line_appender` and `line_appender_header`, and concat them on the end.
I think we can have something like:
```cpp
auto c = l.make_cursor();
l.push_cursor_and_goto(c);
l.append("int x;");
l.pop_cursor();
l.append("x = 233;");
```
or even:
```
auto c = l.make_cursor();
{ // goto header and append
  ScopedCursor _(c);
  l.append("int x;");
} // back to where we stopped
l.append("x = 233;");
```

**Comment by k-ye** [comment]
> FYI I'm using two LineAppender in the C backend: line_appender and line_appender_header, and concat them on the end.

Yep, i saw the OpenGl backend also had similar thing. On Metal, I need to generate the kernel signature **after processing the AST**.

> I think we can have something like:

Not sure if I follow. You said that:

```cpp
auto c = l.make_cursor();   // 1
l.push_cursor_and_goto(c);  // 2
l.append("int x;");
```

Did you mean to append more to `l` between `1` and `2`?

What I meant:

make a cursor, but keep appending to the end (cursor will be in the middle of this stream after appending) --> rewind to the cursor, append from there --> rewind to the end again.

This allows us to rewind to an arbitrary point, not just the head. We can even have multiple cursors.

Basically, the cursor will be an iterator of a linked list.

**Comment by archibate** [comment]
> Did you mean to append more to l between 1 and 2?

Yes, I meant.

> This allows us to rewind to an arbitrary point, not just the head. We can even have multiple cursors.

Is that possible to **go back to the original position** after rewind? i.e. re-rewind.

**Comment by k-ye** [comment]
> Yes, I meant.

Ah ok, then we're on the same page.

> Is that possible to go back to the original position after rewind?

Yep, for example: `rewind_to_end ()` can take you back to the end.
